RSDK-2811 - do not add classify/detect funcs if they are not functional

kim-mishra opened the pull request.

Commit:
Merge branch 'main' of github.com:viamrobotics/rdk

Otterverse committed:
Fix antique-cache docker build (#2266)

edaniels committed:
adjust grpc keepalive (#2267)

kkufieta committed:
RSDK-2029 - Remove slam integration tests (#2264)

penguinland committed:
Rename PWM -> PWM Duty Cycle (#2263)

Commit:
do not add classify/detect funcs if they are not functional

bhaney commented on services/vision/mlvision/ml_model.go:
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
+
 	classifierFunc, err := attemptToBuildClassifier(mlm)
 	if err != nil {
-		logger.Infow("error turning turn ml model into a classifier:", "model", params.ModelName, "error", err)
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
```
Keep this as Infow in order to allow the logger to format the information as it sees fit

bhaney commented on services/vision/mlvision/ml_model.go:
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
 	detectorFunc, err := attemptToBuildDetector(mlm)
 	if err != nil {
-		logger.Infow("error turning turn ml model into a detector:", "model", params.ModelName, "error", err)
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
```
keeps this as Infow

bhaney commented on services/vision/mlvision/ml_model.go:
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
-		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+		detectorFunc, err = checkIfDetectorWorks(ctx, detectorFunc)
+		if err != nil {
+			logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
```
Turn this into Infow

bhaney commented on services/vision/mlvision/ml_model.go:
```diff
@@ -92,9 +105,12 @@ func registerMLModelVisionService(
 	var out []float64
 	me := inMap[name]
+	if me == nil {
+		return nil, errors.New("Invalid input to unpack")
```
add the name into the error as "no such tensor named %q to unpack"

bhaney reviewed (changes requested):
mostly formatting and image related comments

Commit:
addressing PR reviews

Otterverse committed:
[APP-712] Update URLs to use https (#2272)

bhaney commented on services/vision/mlvision/ml_model.go:
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
 	classifierFunc, err := attemptToBuildClassifier(mlm)
 	if err != nil {
-		logger.Infow("error turning turn ml model into a classifier:", "model", params.ModelName, "error", err)
+		logger.Infow("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
```
The way `Infow` works is like a map: You have an initial message, and then a map of messages, like:
```
log.Infow("here is relevant info",
"timestamp", time.Now(),
"name", thing.Name,
"error", err,
)
```

So in your case, you'll want to write something like
`logger.Infow("was not able to turn ml model into classifier", "ml model", params.ModelName, "error", err)`

bhaney commented on services/vision/mlvision/ml_model.go:
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
-		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+		classifierFunc, err = checkIfClassifierWorks(ctx, classifierFunc)
+		if err != nil {
+			logger.Infow("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
```
same here, rather than using Wrapf, use the dict structure of Infow

bhaney commented on services/vision/mlvision/classifier.go:
```diff
@@ -45,9 +45,15 @@ func attemptToBuildClassifier(mlm mlmodel.Service) (classification.Classifier, e
+			return nil, err
+		}
 		if len(probs) == 0 {
-			probs = unpack(outMap, DefaultOutTensorName+"0")
+			probs, err = unpack(outMap, DefaultOutTensorName+"0")
+			if err != nil {
+				return nil, err
+			}
```
Would you ever get to this statement? If probability was not able to be unpacked, wouldn't it just return the error and not come to this statement?

bhaney commented on services/vision/mlvision/classifier.go:
```diff
@@ -62,3 +68,17 @@ func attemptToBuildClassifier(mlm mlmodel.Service) (classification.Classifier, e
+		return nil, errors.New("Nil classifier function")
+	}
+
+	img := &image.RGBA{}
```
This initializes as a (0,0) dimensional image. It's quite an edge case and I wouldn't normally expect this to work. Have you tested this in an integration test? You should add tests to see if these functions behave as expected.

kim-mishra replied:
I tested this with the model that I had been using earlier and it worked, I can write a test as well!

bhaney replied:
I meant in that often there is a "resize" operations that gets done on images when they get fed into an ML model to make them comform to what the model expects. "resizing" an (0,0) image seems like it would have caused a problem, but I guess it didn't!

bhaney commented on services/vision/mlvision/detector.go:
```diff
@@ -51,17 +51,35 @@ func attemptToBuildDetector(mlm mlmodel.Service) (objectdetection.Detector, erro
+		if err != nil {
+			return nil, err
+		}
 		if len(locations) == 0 {
```
if the original "location" is not unpacked, would you ever get to this line?

kim-mishra replied:
I believe so because they are using different keys to get data out of `outputMap`

bhaney commented on services/vision/mlvision/detector.go:
```diff
@@ -83,3 +101,17 @@ func attemptToBuildDetector(mlm mlmodel.Service) (objectdetection.Detector, erro
 	}, nil
 }
+
+func checkIfDetectorWorks(ctx context.Context, df objectdetection.Detector) (objectdetection.Detector, error) {
```
Make a test for this to make sure it works as expected

bhaney reviewed (changes requested):
some more questions, and Infow suggestions

kkufieta committed:
RSDK-2029 - Delete builtin & co (#2271)

DTCurrie committed:
Make TS SDK a peer dependency (#2273)

Commit:
added tests

Commit:
infow updates and tests added

DTCurrie committed:
Disconnect before browser tab unload (#2268)

penguinland committed:
remove comment about digital interrupt callbacks (#2274)

I don't know if we'll ever remove the "servo" type of digital interrupt (see https://viam.atlassian.net/browse/RSDK-1963), but in the meantime this comment is showing up in the [public, auto-generated documentation](https://pkg.go.dev/go.viam.com/rdk/components/board#DigitalInterrupt), and is causing some confusion (per Sierra). Basic digital interrupts have supported callbacks for a long time, and non-basic (servo) digital interrupts probably shouldn't be used.

Eric wrote this in 2021, which I suspect was before the current interface had been fleshed out very well.

Commit:
Merge branch 'main' of github.com:viamrobotics/rdk

bazile-clyde committed:
Find cameras connected to computer after server start (#2276)

stevebriskin committed:
Rsdk-2806: Movement sensor distance to meters (#2275)

bazile-clyde committed:
nit: switch to bool  (#2278)

Commit:
erge branch 'main' of github.com:viamrobotics/rdk

bhaney committed:
RSDK-2224 New vision service (#2254)

Co-authored-by: Khari Jarrett <khari@viam.com>

DTCurrie committed:
Check for self-managed connections (#2279)

Commit:
Merge branch 'main' of github.com:viamrobotics/rdk

Commit:
merge conflicts

Commit:
more merge conflicts

Commit:
fixing some merge conflict issues

Commit:
more merge conflicts fixes
